test/test_jit.py

```diff
@@ -7069,6 +7089,95 @@ def test_dtype_op_shape2(self):
 
         self._test_dtype_op_shape(ops, args=[1], input_dims=4)
 
+
+    def _test_binary_op_shape(self, ops, input_dims=1):
+
+        dtypes = [torch.float32, torch.float64, torch.int64, torch.int32, torch.bool]
+
+        if input_dims == 0:
+            shape = '1'
+        else:
+            shape = '[' + ('1,' * 4) + ']'
+            for _ in range(1, input_dims):
+                shape = '[' + ",".join([shape] * 4) + ']'
+
+        template = dedent('''
+        def func():
+            arg1 = {}
+            arg2 = {}
+            return torch.{}(arg1, arg2)
+        ''')
+
+        args = []
+        for dtype in dtypes:
+            args = args + ["torch.tensor({}, dtype={})".format(shape, dtype)]
+        args = args + [1, 1.5]
+
+        def isBool(arg):
+            return type(arg) == bool or (type(arg) == str and "torch.bool" in arg)
+
+        for op in ops:
+            for first_arg in args:
+                for second_arg in args:
+                    # subtract not supported for bool
+                    if (op == 'sub' or op == 'div') and (isBool(first_arg) or isBool(second_arg)):
+                        continue
+                    # div not implemneted correctly for mixed-type scalar params
+                    if (op == 'div' and type(first_arg) != type(second_arg) and type(first_arg) != str and type(second_arg) != str):
+                        continue
+                    return_line = "torch.{}({}, {})".format(op, first_arg, second_arg)
+                    # uncomment for debugging a failed test:
+                    # print("testing {}".format(return_line))
+                    code = template.format(first_arg, second_arg, op)
+                    scope = {}
+                    exec(code, globals(), scope)
+                    non_jit_result = scope['func']()
+
+                    cu = torch.jit.CompilationUnit(code)
+                    graph = cu.func.graph
+                    torch._C._jit_pass_complete_shape_analysis(graph, (), False)
+                    # use dim=-1 to represent a python/jit scalar.
+                    dim = -1 if type(first_arg) != str and type(second_arg) != str else non_jit_result.dim()
+                    dtype = non_jit_result.dtype
```

Comment:
I didn't comment when you used a similar pattern in the last PR when I should have, but regardless:

I think the testing logic here would be simplified if you checked the properties of the output type against the non_jit_result type directly instead of constructing a string to match against.
```
inferred_out_type = next(graph.outputs()).type()
self.assertTrue(non_jit_result.scalarType() = non_jit_result.dtype)
self.assertTrue(inferred_out_type.dim() = out_val.dim)
```

Reply:
hmm.. I could do something like that for at least some cases, but not all inferred types have a scalarType(). e.g. for `torch.mul(1, 1)` the inferred output is: `3 defined in (%3 : int = aten::mul(%arg1.1, %arg1.1) # <string>:4:12`. 

I suppose this is a bug, but `inferred_output.scalarType()` in that case throws 
`RuntimeError: false INTERNAL ASSERT FAILED at ../aten/src/ATen/core/jit_type.h:542, please report a bug to PyTorch. Expected a tensor type`

Reply:
Yes it expects the input to be a tensor by design. Why are you testing `torch.mul(1, 1)`? You did not touch that code in shape analysis

Reply:
> Yes it expects the input to be a tensor by design. Why are you testing `torch.mul(1, 1)`? You did not touch that code in shape analysis

I believe the first arg gets converted to a wrapped tensor and handled by the `mul(Tensor self, Scalar other)` handler? Regardless, shouldn't we have tests that confirm the result is the same dtype between python and jit?

Reply:
There's no need to confirm the type of `torch.mul(1, 1)` that is already statically typechecked to be an int.

Reply:
I just iterate through and check every combination. The behavior has actually changed in some cases.

On master:
```
>>> torch.set_default_dtype(torch.float)
>>> torch.mul(1, 1.0).dtype
torch.float64
```
On this branch:
```
>>> torch.set_default_dtype(torch.float)
>>> torch.mul(1, 1.0).dtype
torch.float32
```
Without the line 
`auto default_dtype = at::typeMetaToScalarType(caffe2::get_default_dtype());` 
in shape_analysis.cpp this fails, so I think the test is valid. I was just verifying and accidentally committed, hence the commented out code you noticed in the review.